SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,519 @@
+//===--- AccessUtils.swift - Utilities for analyzing memory accesses ------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+// This file provides a set of utilities for analyzing memory accesses.
+// It defines the following concepts
+// - `AccessBase`: represents the base address of a memory access.
+// - `AccessPath`: a pair of an `AccessBase` and `SmallProjectionPath` with the
+//   the path describing the specific address (in terms of projections) of the
+//   access.
+// - `AccessStoragePath`: identifies the reference (or a value which contains a
+//   reference) an address originates from.
+//===----------------------------------------------------------------------===//
+import SIL
+
+/// "AccessBase describes the base address of a memory access (e.g. of a `load` or `store``).
+/// The access address is either the base address directly or an address
+/// projection of it.
+/// The following snippets show examples of memory accesses and their respective bases.
+///
+/// ```
```

Comment:
I would rather put this into the top-level file comment, because it shows the relation of all utilities defined in this file